src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -2296,105 +2296,117 @@ private void handleStateNormal(final InetAddress endpoint, final String status)
         // Order Matters, TM.updateHostID() should be called before TM.updateNormalToken(), (see CASSANDRA-4300).
         UUID hostId = Gossiper.instance.getHostId(endpoint);
         InetAddress existing = tokenMetadata.getEndpointForHostId(hostId);
-        if (replacing && isReplacingSameAddress() && Gossiper.instance.getEndpointStateForEndpoint(DatabaseDescriptor.getReplaceAddress()) != null
-            && (hostId.equals(Gossiper.instance.getHostId(DatabaseDescriptor.getReplaceAddress()))))
-            logger.warn("Not updating token metadata for {} because I am replacing it", endpoint);
-        else
-        {
-            if (existing != null && !existing.equals(endpoint))
+        boolean acquiredTokenLock = false;
+        try {
+            if (replacing && isReplacingSameAddress() && Gossiper.instance.getEndpointStateForEndpoint(DatabaseDescriptor.getReplaceAddress()) != null
+                && (hostId.equals(Gossiper.instance.getHostId(DatabaseDescriptor.getReplaceAddress()))))
+                logger.warn("Not updating token metadata for {} because I am replacing it", endpoint);
+            else
+            {
+                if (existing != null && !existing.equals(endpoint))
+                {
+                    // It's safe for us to say we acquired the lock before we've, as this is simply used for calling unlock.
+                    // Unlock will throw if a separate thread that is not holding it tries to unlock it.
+                    acquiredTokenLock = true;
+                    tokenMetadata.lock();
```

Comment:
Seems odd to flip the flag before we actually lock
```suggestion

                    tokenMetadata.lock();
                    acquiredTokenLock = true;
```

Reply:
Yup so there is a trade-off here: If we acquire _after_ we lock, there is a chance that we do not reach the `acquiredTokenLock` statement (such as an interrupt), which places us into a state of deadlock. Whereas if we assign the lock _before_ we acquire it, the worst that can happen is we throw an exception when we try to unlock the lock as our current thread doesn't own it. And seeing as an exception has already been thrown in this case (i.e. lock() failed), it should be relatively harmless to throw yet _another_ exception.